### PR TITLE
Create Long Haul Yaml

### PR DESCRIPTION
**Pull request by philipktlin**
Create a YAML file for Long Haul test; and need to generate a new artifact info file; which will contains artifact build number.  This is required to use pulling docker images for long haul test.

Already tested images.yaml and partially tested longhaul.yaml.  Will create a new build for long haul once this is committed.

### CODE REVIEW

File: builds/e2e/longhaul.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+trigger: none
+
+jobs:
+
+################################################################################
+  - job: LinuxAMD64Moby
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: This is a huge nit, but all of the other job names are in snake_case
```suggestion
  - job: linux_amd64_moby
```

File: builds/e2e/longhaul.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+################################################################################
+    displayName: Linux AMD64 Moby
+    pool:
+      name: Azure-IoT-Edge-Core
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: I would check but I think we typically set the pool name via a build variable

File: builds/e2e/longhaul.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        - run-long-haul -equals true
+        - Agent.OS -equals Linux
+        - Agent.OSArchitecture -equals X64
+    steps:
+      - task: DownloadBuildArtifacts@0
+        displayName: 'Download Iotedged Artifacts'
+        inputs:
+          buildType: specific
+          project: 'b32aa71e-8ed2-41b2-9d77-5bc261222004'
+          pipeline: 55463
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: We should make the `project`, `pipeline`, and `artifactName` fields as variables.

File: builds/e2e/longhaul.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      - task: DownloadBuildArtifacts@0
+        displayName: 'Download Core-Linux Artifacts'
+        inputs:
+          buildType: specific
+          project: 'b32aa71e-8ed2-41b2-9d77-5bc261222004'
+          pipeline: 55174
+          buildVersionToDownload: latestFromBranch
+          artifactName: 'core-linux'
+          itemPattern: |
+           core-linux/IotEdgeQuickstart.linux-x64.tar.gz
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: These should be indented by another space (2 total)

File: builds/e2e/longhaul.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+           core-linux/e2e_deployment_files/long_haul_deployment.template.json
+           core-linux/scripts/linux/runE2ETest.sh
+      - task: AzureKeyVault@1
+        displayName: 'Azure Key Vault: edgebuildkv'
+        inputs:
+          azureSubscription: 'azureiotedge-arm'
+          KeyVaultName: edgebuildkv
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: `azureSubscription` and `KeyVaultName` needs to be variables.

File: builds/e2e/longhaul.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+          SourceFolder: '$(System.ArtifactsDirectory)/core-linux'
+          TargetFolder: '$(Agent.HomeDirectory)/../artifacts/core-linux'
+          CleanTargetFolder: true
+      - task: Bash@3
+        displayName: 'Run long haul deployment'
+        inputs:
+          targetType: inline
+          script: |
+            declare -a cnreg=( $(edgebuilds-azurecr-io) )
+            sudo chmod +x $(Agent.HomeDirectory)/../artifacts/core-linux/scripts/linux/runE2ETest.sh
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]: You shouldn't need to set the mode to `+x`. `git` should be able to preserve the mode. You might need to jump on a linux machine to fix it and check it in.

File: builds/e2e/longhaul.yaml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+          SourceFolder: '$(System.ArtifactsDirectory)/iotedged-ubuntu-armhf'
+          TargetFolder: '$(Agent.HomeDirectory)/../artifacts/iotedged-ubuntu-armhf'
+          CleanTargetFolder: true
+      - task: CopyFiles@2
+        displayName: 'Copy core-linux artifact'
+        inputs:
+          SourceFolder: '$(System.ArtifactsDirectory)/core-linux'
+          TargetFolder: '$(Agent.HomeDirectory)/../artifacts/core-linux'
+          CleanTargetFolder: true
+      - task: Bash@3
```
Comment (philipktlin) [inline builds/e2e/longhaul.yaml]: Should I make this step as template?

File: scripts/linux/createArtifactInfo.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+# This script is used to create artifact info file.
+
+usage()
+{
+    echo "$SCRIPT_NAME [options]"
+    echo ""
+    echo "options"
+    echo " -outputFolder        output folder for artifact info file"
```
Comment (myagley) [inline scripts/linux/createArtifactInfo.sh]: [nit] linux "long name" arguments typically use two dashes `--` and are `kebab-case`. For example:

```
echo " --output-folder         output folder for artifact info file"
echo " --build-number          build number of artifact"
```

File: builds/e2e/longhaul.yaml (hunk 6 above)
Comment (myagley) [reply builds/e2e/longhaul.yaml]: It looks like the mode was changed on this file from `0644` to `0755` below. You may not need this line.

File: builds/e2e/longhaul.yaml (hunk 1 above)
Comment (philipktlin) [reply builds/e2e/longhaul.yaml]: I don't know I can use suggestions; love it!

File: builds/e2e/longhaul.yaml (hunk 2 above)
Comment (philipktlin) [reply builds/e2e/longhaul.yaml]: No.  I checked existing yamls; but use variable for windows pipeline in ci\dotnet.yaml only.  I will keep it simple for now.

File: builds/e2e/longhaul.yaml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        - Agent.OS -equals Linux
+        - Agent.OSArchitecture -equals ARM
+    variables:
+      edgelet.artifact.name: 'iotedged-ubuntu-armhf'
+    steps:
+      - task: DownloadBuildArtifacts@0
+        displayName: 'Download Edgelet Artifacts'
+        inputs:
+          buildType: specific
+          project: $(mszure.one.project)
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]:
```suggestion
          project: $(vsts.project)
```

File: builds/e2e/longhaul.yaml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+          pipeline: $(edgelet.package.build)
+          buildVersionToDownload: latestFromBranch
+          artifactName: $(edgelet.artifact.name)
+      - task: DownloadBuildArtifacts@0
+        displayName: 'Download Images Artifacts'
+        inputs:
+          buildType: specific
+          project: $(mszure.one.project)
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]:
```suggestion
          project: $(vsts.project)
```

File: builds/e2e/longhaul.yaml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (myagley) [inline builds/e2e/longhaul.yaml]:
```suggestion
          project: $(vsts.project)
```

File: builds/e2e/longhaul.yaml [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        - Agent.OS -equals Linux
+        - Agent.OSArchitecture -equals X64
+    variables:
+      edgelet.artifact.name: 'iotedged-ubuntu-amd64'
+    steps:
+      - task: DownloadBuildArtifacts@0
+        displayName: 'Download Edgelet Artifacts'
+        inputs:
+          buildType: specific
+          project: $(mszure.one.project)
```
[the remark this answers is not in the document]
Comment (myagley) [reply builds/e2e/longhaul.yaml]: `msazure.one` is an internal name. We shouldn't make it public.

[the remark this answers is not in the document]
Comment (myagley) [reply builds/e2e/longhaul.yaml]: You might be able to use the `System.TeamProject` or `System.TeamProjectId` variables here: https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables

[the remark this answers is not in the document]
Comment (philipktlin) [reply builds/e2e/longhaul.yaml]: Thanks. good to know this predefined variable.

**Pull request merged**